### DISCUSSION TITLE
Pendant MPG DRO wireless for grblHAL CNC

### DISCUSSION THREAD

**Discussion by stevempotter** [terjeio/grblHAL#313] [Ideas]
I am planning to build a wireless pendant for my CNC (MPCNC Primo, controlled by Phil Barrett's Teensy 4.x Breakout Board).
I think that to use a pair of ESP32s and their ESP-NOW functionality will work well, based on some initial tests. 

If you are not familiar with ESP-NOW, it is a low-latency proprietary communications protocol intended for controlling smart lighting and whatnot. It connects instantly and stays connected (paired) as long as the devices are powered on and in range (>100m), and each message can carry 250 bytes. They can power up and reconnect in milliseconds when switched on. Com can be encrypted if desired. All that is needed to pair devices is their MAC address. It does not use Wifi, but does use the 2.4GHz frequency range. You can use any combination of bidirectional broadcasting or narrowcasting, one-to-one, one-to-many (up to 20 devices).  Here is more info: [https://www.espressif.com/en/products/software/esp-now/resources](https://www.espressif.com/en/products/software/esp-now/resources)
Here are the ESP-32 boards I am using: [https://www.tinytronics.nl/shop/en/platforms/ttgo/lilygo-ttgo-t-display-v1.1-esp32-with-1.14-inch-tft-display](https://www.tinytronics.nl/shop/en/platforms/ttgo/lilygo-ttgo-t-display-v1.1-esp32-with-1.14-inch-tft-display)

My question is: if we call the ESP-32 in the pendant the Sender and the one in the grblHAL box the Receiver, how do I go about interfacing a high-spec MCU like the ESP-32 to the Breakout Board running grblHAL? There will also be info going the other way (from Receiver to Sender), such as the current coordinates for the small LCD on the ESP-32 board. Do I use its ethernet connection, I2C, USB, or something else? Does the firmware already have hooks to take commands from pendants with lots of knobs and buttons (not a touchscreen or just a keyboard/keypad). Raised on old Tektronix oscilloscopes, I am a knob and button lover. Much preferred over touchscreen or mouse-controlled interface on a computer, or complex menu hierarchy to click through with one encoder button. I will be using a 1000 pulses/rev encoder for the MPG, and also an RC joystick with Hall effect sensors. 

Any advice on how to get started on the interfacing will be appreciated!

**Comment by phil-barrett** [comment]
Take a look at the [I2C keypad code](https://github.com/grblHAL/Plugin_I2C_keypad).  It might be possible to just generate the codes that the plugin expects.  I'd be happy to help out with the hardware side of things as this is a topic near and dear to me. I have several designs that use the I2C header that could easily be adapted.   

Perhaps this topic should be moved to grblHAL/Plugin_I2C_keypad though there isn't a discussions section.  Terje, anything against enabling that?

**Reply by stevempotter** [reply]
I2C sounds good. Phil, I would love to see your several designs to help move my project along in the most sensible way based on your experience. I prefer not to use a keypad approach, but to transmit raw pulses (for proportional speed of jogging) and switch states. Is that possible? Long ago, I designed such a system to move a microscope stage with rotary optical encoders, programmed in LabView, and to have the movement be precise, no-latency, and proportional was wonderful. Same goes for another long-time hobby of mine, RC plane flying.

**Reply by phil-barrett** [reply]
The Keypad plug-in essentially uses key codes but you can generate them any way you want.  There are many ways to tell grblHAL where and how to move. I'd start with the I2C receiver pumping out the key codes to get the ESP32 side working and move from there. The code is fairly easy to understand, iirc.  But, Terje will probably have better ideas on how to proceed.

I don't have a github site for this stuff.  May create a quick-n-dirty one in a bit. Will post when done.

**Reply by terjeio** [reply]
> Perhaps this topic should be moved to grblHAL/Plugin_I2C_keypad though there isn't a discussions section. Terje, anything against enabling that?

No, discussions should be added to all repositories. I do not believe it is possible to move the discussion there.

The I2C keypad interface is good for buttons, including buttons for jogging. I am not sure that adding a MPG interface to that is the best way forward. My [MPG & DRO](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack) uses the I2C interface for the buttons and a secondary UART stream for the MPG encoders, the latter by sending jog commands.

MPG handling is IMO a bit tricky - should jog commands be issued or should the encoder somehow send pulses directly to the motors? If the latter then there is the problem of scaling, e.g. my lathe is configured for around 3000 pulses/mm and my CO2 laser for around 100. With jog commands the issue is how to move smoothly - and two modes are possible, velocity or position based. Rather raw implementations for these can be found in the [encoder plugin](https://github.com/grblHAL/Plugin_encoder/blob/master/encoder.c). I use position based for my MPG & DRO, an issue is jerky movements due to the encoders having detents - impossible to move smoothly at low speeds. I just got a 600 PPR encoder with no detents, at some point I want to play with that - likely with my RP2040 version of the MPG &DRO...

Interfacing with the controller could also be done using the ModBus protocol, there are at least one [user tinkering with that](https://github.com/terjeio/grblHAL/discussions/253).

**Reply by phil-barrett** [reply]
Perhaps I am looking at this too simplistically but an MPG interface is similar to the sender interface for jogging. Turning the wheel 10 clicks should advance the selected axis by 10 units worth of moves. The units would be a user selection, just like in the sender. Whether that is 10 messages or 1 message with a 10X factor is an optimization question, I believe.  Clearly, this requires some sort of processor. In the ATTiny design I put up, the microcontroller reads the quad encoder + units switches and formats up messages that get sent over I2C. I have ATTiny code that does all that but have not done the grblHAL interface. My thinking was to emulate the Keypad interface at first.  I put it aside thinking USB is a better way to go. 

> should the encoder somehow send pulses directly to the motors?

I think not. Scaling is certainly an issue but would it not make more sense to have an interface (in the core?) that allows any manner of connection and device to be used?  Game controller key pads, analog joysticks, rotary hand wheel encoders, pre-programmed buttons, ...   I imagine that this could be used for stand alone control panels similar to the [DDCS](https://www.amazon.com/DDCSV3-1-Upgrade-Stand-alone-Controller-Engraver/dp/B07SR854D1).

Whether the interface is Async/UART or I2C (or even SPI) is worth discussing though it is somewhat orthogonal to how to interface to the grbl core.  I can see the advantage of wired RS485 (ie Modbus) for reliability.  For an ESP32/Wifi or BT connection, perhaps I2C is better? 

Another interface to consider is Hosted USB on the grblHAL breakout board.  While this is only on one (as yet unreleased) board it has some benefit as well.

**Reply by terjeio** [reply]
> Perhaps I am looking at this too simplistically but an MPG interface is similar to the sender interface for jogging. Turning the wheel 10 clicks should advance the selected axis by 10 units worth of moves.

Yes, that is a bit on the simple side. With keyboard/keypad jogging the sample rate is rather slow (per keypress) and the feed rate is predetermined. Unless delaying moves until the encoder is at rest sampling has to be done while it is turning and commands issued to match, this includes changing the feedrate in sync with the turning speed.

>>  should the encoder somehow send pulses directly to the motors?

> I think not. Scaling is certainly an issue but would it not make more sense to have an interface (in the core?) 

Sure - it has to be done via a core interface somehow, by directly I meant bypassing the parser and the planner.

There are three ways of controlling the axes?

* Via buttons/keys - easy to implement, done and dusted?
* Via encoder input that emits jog commands (gcode). No so easy to get right? Initial attempts can be found in the encoder plugin.
* Via encoder input that controls the motors by bypassing the parser and the planner stages. Not easy either as scaling is required. A software PLL is possible?

Many different interfaces can be employed to supply the data needed for these.

**Reply by phil-barrett** [reply]
> Via encoder input that emits jog commands (gcode). No so easy to get right? Initial attempts can be found in the encoder plugin.

Ultimately, I think this is the way to go for a jog wheel.  Yes, wheel clicks are faster than button presses but they can be accumulated over a relatively short period of time (say, 100 mS?) so each click of the wheel does not (necessarily) mean one message. That is what I did with the ATTiny code. The physical axis movement can lag even slow button presses so I don't see it as a responsiveness issue.  Perhaps the first click needs to happen faster to feel more responsive. A fair amount of tuning will be needed to make it smooth, of course. Perhaps that is what you mean by "get right".

**Reply by stevempotter** [reply]
Phil and Terjeio, I am glad to see this excellent discussion! One useful piece of info I measured today: the ESP-NOW messages (16 byte payload in my test) have a latency of 1.1 to 1.3 milliseconds. That will allow at least 500Hz of messages to be sent from the pendant. I agree that the sender ought to do some processing to accumulate pulses or measure velocity before sending a command, as most people will not have an encoder with so many pulses per revolution (mine, soon to arrive, is 1000ppr with no detents, made for a servomotor shaft).

**Reply by phil-barrett** [reply]
I have not looked at ESP-NOW but I wonder about error handling and noise immunity.  A machine environment can have a lot of EMI so it is possible that some packets could get lost?  TCP/IP has resend.  Is there an overview you could point me to?

The MPG I linked to above has 100 PPR with detents.  Not sure how "twitchy" 1000 PPR and no detents will be.

**Reply by stevempotter** [reply]
RE: ESP-NOW: The link I pointed to in my original post is a good place to start, or if you prefer a video, see this one by Rui Santos of Random Nerd Tutorials: [https://youtu.be/qxwXwNS3Avw](https://youtu.be/qxwXwNS3Avw) . Yes, you can have every message sent confirmed by the receiver and program a timeout and a resend-until-succeed into your code. I suppose you could even echo back the sent message, and compare them before sending an "execute" message, to be extra careful. The max payload is 250 bytes per message, plenty for a pendant, especially if you are sending hundreds of messages per second. I have not played around with the option to encrypt messages, but I imagine that may also improve your confidence that the right message gets received. I have not done a range test myself, but the Random Nerd Tutorials guy did and found it was >250 meters (open field). If one of the ESP-32s has a reset, it takes only a few ms to reboot and get the messages going again, unlike any other wireless com I know of.

**Reply by stevempotter** [reply]
Terjeio, your MPG & DRO is beautiful! It looks like it has all the functionality I was planning on having. Why did you use a secondary UART stream for the encoders? I  am hoping I can pump everything needed into the I2C port. I am not too familiar with gCode; Does grblHAL use the same commands as regular grbl, as described here? [https://github.com/gnea/grbl/wiki/Grbl-v1.1-Jogging](https://github.com/gnea/grbl/wiki/Grbl-v1.1-Jogging)
That was a couple of years ago...Have things advanced since then in terms of proportional pendant/joystick implementations?

**Reply by phil-barrett** [reply]
OK. Sound like a plan. You could probably just use a simple check sum to verify each packet is ok.  Not 100% but sufficient, especially if you use a 16 bit or 32 bit check size. 

That video was mercifully short (though his base boost a bit too much).  I like it when they get straight to it without screwing around.

Does NOW work with the ESP-12F. I've been playing with a them.

**Reply by terjeio** [reply]
>  A fair amount of tuning will be needed to make it smooth, of course. Perhaps that is what you mean by "get right".

Yes, that is what I mean.

> Why did you use a secondary UART stream for the encoders?

It just happened that way? I wrote the I2C keypad handler for my CO2 laser and reused that for the MPG & DRO. And since that has an inbuilt gcode sender I found it easier to use a UART for streaming the gcode. Using I2C for two way communication and variable length messages is IMO harder to get right than just adding another UART stream.

>  Does grblHAL use the same commands as regular grbl, as described here?

Yes, the only "change" is that the infamous jog cancel bug is fixed so that long distance jog commands can be issued and safely cancelled.

> Have things advanced since then in terms of proportional pendant/joystick implementations?

I believe the current interface can be used, but improvements are needed on the sending side. Phil seconded that above. So no, nothing has changed since that interface was added.

---

For the RP2040 port of the MPG & DRO code (I am HALifying that too to make it easier to port) I am considering using ModBus instead of I2C/plain UART for comms. I am also going to look into using ESP Bluetooth and/or Wifi modules as an option instead of wired connections, but I am not going to put the bulk of the code on them unless they have a proper debug interface.

**Reply by stevempotter** [reply]
"Does NOW work with the ESP-12F. I've been playing with a them." [I can't figure out how to quote in this forum]
I believe that any ESP32 or ESP8266 has ESP-NOW capability
Yes, Rui Santos of RNT is my hero. Here is more info about ESP-NOW on ESP8266:
[https://randomnerdtutorials.com/esp-now-esp8266-nodemcu-arduino-ide/](https://randomnerdtutorials.com/esp-now-esp8266-nodemcu-arduino-ide/)

**Reply by phil-barrett** [reply]
> [I can't figure out how to quote in this forum]

![image](https://user-images.githubusercontent.com/16570958/121097600-14f60500-c7a9-11eb-95e1-ef0a281d9064.png)

**Reply by stevempotter** [reply]
>> Have things advanced since then in terms of proportional pendant/joystick implementations?

> I believe the current interface can be used, but improvements are needed on the sending side. Phil seconded that above. So no, nothing has changed since that interface was added.

The grbl jogging page says:

> Flash and memory restrictions prevent the original envisioned implementation...

Since we are using much better microprocessors in grblHAL than an Arduino with grbl, doesn't that imply that such limitations no longer apply and we can dream big about how to implement pendants?

**Reply by terjeio** [reply]
> Since we are using much better microprocessors in grblHAL than an Arduino with grbl, doesn't that imply that such limitations no longer apply and we can dream big about how to implement pendants?

Ok, I should have read that page again. I cannot remember seeing the provided formulas before, interesting.

Pendants can be implemented in grblHAL by adding plugins or driver extensions that submit jog commands to the core, the keypad and encoder plugins does that already. Pendands may simulate keypresses and or/use analog inputs and/or pulse signals, they might be connected via GPIO pins, USB,  WiFi, Bluetooth and what not. So IMO limits are imposed by availablility of interfaces and/or pins in the MCU, not the core code. Stop dreaming and start writing code? You may even add pendant support without changing a single line of downloaded code...

**Reply by stevempotter** [reply]
OK, I am well along with the coding. Today I wrote the code that made bidirectional ESP-NOW comm very reliable on my two ESP-32s. I have transmitted over one million messages at over 400 messages/second without a single error so far. That is with the two devices 28 feet apart, with the sender on a Lipo battery, as it will be in the pendant. The trick was to make sure that the sender waits for its message to be echoed back to it from the receiver before sending its next one. When I do that, with no delay statements in the code, I am getting an average of 2.4ms per round-trip loop. That is with a small 8-byte message. It will take longer to send larger messages -- I will measure that at some point, and also the effect of encrypting the messages. But this has convinced me that ESP-NOW is ideal for wireless pendant signals where responsiveness and reliability is desired.

Getting the receiving ESP-32 to send gcode commands out its I2C lines should be straightforward, right?
Now I need to read up to understand grbl (which I have not used yet, I am using Marlin so far) and what is the difference between a plugin and a driver extension.

**Reply by terjeio** [reply]
> Getting the receiving ESP-32 to send gcode commands out its I2C lines should be straightforward, right?

I guess so, but you have to make some kind protocol for that and add it to the grblHAL driver side as well. The keypad plugin can serve as an example for a simple protocol. Receiving the messages should be done in the background (via interrupts), submitting the gcode for processing should take place in the foreground process. 

> what is the difference between a plugin and a driver extension.

A plugin is not bound to a specific processor/driver, any hardware accesses has to go through the HAL or another well defined API. A driver extension access MCU specific peripherals directly and cannot be shared with other drivers without changes.

**Reply by stevempotter** [reply]
> most people will not have an encoder with so many pulses per revolution (mine, soon to arrive, is 1000ppr with no detents, made for a servomotor shaft).

Great news! I have been playing with my 1024 ppr rotary optical encoder (OMRON E6B2-CWZ6C 05421M) and a library that makes use of the [ESP32's built in pulse counters](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/pcnt.html#:~:text=The%20PCNT%20(Pulse%20Counter)%20module,increment%20or%20decrement%20the%20counter.) that do not require the use of interrupts. The library #include <ESP32Encoder.h> works like a charm! And if I use the full quadrature option (ROE.attachFullQuad(ROE_B, ROE_A);) I get 4096 counts per revolution from my 1024 ppr device! That means that even with very fast sampling (say, 2ms/sample) I usually get several tens of counts added or subtracted to the current count. This will allow precise jogging even at slow knob-turning rates.

**Comment by phil-barrett** [comment]
I've placed 3 kicad projects on [I2C Headers](https://github.com/phil-barrett/I2C_Headers).

**Reply by stevempotter** [reply]
Thanks, Phil, I will look with keen interest. BTW, what are Hlt and St pins in the I2C bus on your board? I see some mention of Halt and Strobe in your manual but have no idea what those are for in regard to I2C interfaces.

**Reply by phil-barrett** [reply]
Hlt is a pass-through to the grblHAL reset input on the Teensy.  The idea is that if an MPG has an EStop button, you want that to go straight to the processor with nothing else in the way.

Stb is the strobe function and is used for signaling between the two devices (grblHAL processor and MPG/Key Pad/whatever).

**Reply by terjeio** [reply]
> Stb is the strobe function and is used for signaling between the two devices 

Yep, it is used as signal events (key presses as of now) and is active low. The device attached should use an open drain (OD) pin for signalling so that multiple devices can request attention (by wire OR'ing).
